spyder/plugins/ipythonconsole/widgets/kernel_connection_manager.py

```diff
@@ -0,0 +1,409 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © Spyder Project Contributors
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+"""Kernel connection manager."""
+
+# Standard library imports
+import os
+import os.path as osp
+import re
+from threading import Lock
+import uuid
+
+# Third-party imports
+from jupyter_core.paths import jupyter_runtime_dir
+from qtpy.QtCore import QThread
+from zmq.ssh import tunnel as zmqtunnel
+
+# Local imports
+from spyder.api.translations import get_translation
+from spyder.plugins.ipythonconsole.utils.manager import SpyderKernelManager
+from spyder.plugins.ipythonconsole.utils.client import SpyderKernelClient
+from spyder.plugins.ipythonconsole.utils.ssh import openssh_tunnel
+from spyder.plugins.ipythonconsole.utils.stdfile import StdFile
+
+
+# Localization
+_ = get_translation("spyder")
+PERMISSION_ERROR_MSG = _(
+    "The directory {} is not writable and it is "
+    "required to create IPython consoles. Please "
+    "make it writable."
+)
+
+if os.name == "nt":
+    ssh_tunnel = zmqtunnel.paramiko_tunnel
+else:
+    ssh_tunnel = openssh_tunnel
+
+
+class KernelConnection:
```

Comment:
This is not a widget, so it doesn't belong in `ipythonconsole/widgets`. Please move it to `ipythonconsole/widgets/utils` instead.

Besides, this class not only stores connection info but it's in charge of shutting down the kernel too. Perhaps `KernelHandler` would be a better name? (`KernelManager` would be ideal but it's already taken by the Jupyter architecture).

If you think that's a good idea, then I think the file that contains it should be named `kernel_handler.py`.